packages/flutter_tools/lib/src/build_runner/resident_web_runner.dart

```diff
@@ -404,6 +453,130 @@ class ResidentWebRunner extends ResidentRunner {
+      packagesFilePath,
+    );
+    await device.devFS.create();
+    await _updateDevFS(fullRestart: true);
+    device.generator.accept();
+    await device.device.startApp(applicationPackage,
+      mainPath: target,
+      debuggingOptions: debuggingOptions,
+      platformArgs: <String, Object>{
+        'uri': 'http://$effectiveHostname:$hostPort',
+      },
+    );
+    return attach(
+      connectionInfoCompleter: connectionInfoCompleter,
+      appStartedCompleter: appStartedCompleter,
+    );
+  }
+
+  /// Different implementations for the incremental compiler based workflow.
+  Future<int> _attachExperimental({
+    Completer<DebugConnectionInfo> connectionInfoCompleter,
+    Completer<void> appStartedCompleter,
+  }) async {
+    final Chrome chrome = await ChromeLauncher.connectedInstance;
+    final ChromeTab chromeTab = await chrome.chromeConnection.getTab((ChromeTab chromeTab) {
+      return chromeTab.url.contains(debuggingOptions.hostname);
+    });
+    _wipConnection = await chromeTab.connect();
+    appStartedCompleter?.complete();
+    connectionInfoCompleter?.complete();
+    final int result = await waitForAppToFinish();
+    await cleanupAtFinish();
+    return result;
+  }
+
+  Future<OperationResult> _restartExperimental({
+    bool fullRestart = false,
+    bool pauseAfterRestart = false,
+    String reason,
+    bool benchmarkMode = false,
+  }) async {
+    final UpdateFSReport report = await _updateDevFS(fullRestart: fullRestart);
+    if (report.success) {
+      device.generator.accept();
+    } else {
+      await device.generator.reject();
+    }
+    final String modules = report.invalidatedModules
+      .map((String module) => '"$module"').join(',');
+    if (fullRestart) {
+      await _wipConnection.sendCommand('Page.reload');
+    } else {
+      try {
+        await _wipConnection.debugger
+          .sendCommand('Runtime.evaluate', params: <String, Object>{
+          'expression': 'window.\$hotReloadHook([$modules])',
+          'awaitPromise': true,
+          'returnByValue': true,
+        });
+      } on WipError catch (err) {
+        printError(err.toString());
+        return OperationResult(1, '');
+      }
+    }
+    return OperationResult.ok;
+  }
+
+  Future<UpdateFSReport> _updateDevFS({ bool fullRestart = false }) async {
+    final bool isFirstUpload = !assetBundle.wasBuiltOnce();
+    final bool rebuildBundle = assetBundle.needsBuild();
+    if (rebuildBundle) {
+      printTrace('Updating assets');
+      final int result = await assetBundle.build();
+      if (result != 0) {
+        return UpdateFSReport(success: false);
+      }
+    }
+    final List<Uri> invalidatedFiles = await ProjectFileInvalidator.findInvalidated(
+      lastCompiled: device.devFS.lastCompiled,
+      urisToMonitor: device.devFS.sources,
+      packagesPath: packagesFilePath,
+    );
+    final Status devFSStatus = logger.startProgress(
+      'Syncing files to device ${device.device.name}...',
```

Comment:
Since this is Web-specific code, maybe refer to the "browser" rather than "device"?

Reply:
Or just remove "device"? Should still be clear: "Syncing files to chrome"

Reply:
Might be "Syncing files to Edge" someday